spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestDataFrameWriterV2.java

```diff
@@ -244,4 +247,132 @@ public void testMergeSchemaSparkConfiguration() throws Exception {
             row(4L, "d", 140000.56F)),
         sql("select * from %s order by id", tableName));
   }
+
+  @TestTemplate
+  public void testMergeSchemaIgnoreCastingLongToInt() throws Exception {
+    sql(
+        "ALTER TABLE %s SET TBLPROPERTIES ('%s'='true')",
+        tableName, TableProperties.SPARK_WRITE_ACCEPT_ANY_SCHEMA);
+
+    Dataset<Row> bigintDF =
+        jsonToDF(
+            "id bigint, data string",
+            "{ \"id\": 1, \"data\": \"a\" }",
+            "{ \"id\": 2, \"data\": \"b\" }");
+
+    bigintDF.writeTo(tableName).append();
+
+    assertEquals(
+        "Should have initial rows with long column",
+        ImmutableList.of(row(1L, "a"), row(2L, "b")),
+        sql("select * from %s order by id", tableName));
+
+    Dataset<Row> intDF =
+        jsonToDF(
+            "id int, data string",
+            "{ \"id\": 3, \"data\": \"c\" }",
+            "{ \"id\": 4, \"data\": \"d\" }");
+
+    assertThatCode(() -> intDF.writeTo(tableName).option("merge-schema", "true").append())
```

Comment:
the test seems to pass independent of whether this is set or not

Reply:
nvm. the test fails when `merge-schema` is set to `true` but your fix isn't applied. I think it would make sense to add a comment to make this clearer in this test

Reply:
@rocco408 could you please add a comment before this line that explains why we're testing with `merge-schema=true`